### PR TITLE
Tools/Fetch/Git: add `commit` parameter to support specific tags/hashes

### LINKED ISSUE #924: Git plugin should (better) support downloading specific tags or commit hashes
From what I can see, the Fetch-Tool plugin for Git features a `branch` option, but no dedicated way to specify a single tag or commit hash.
Putting a tag name into the `branch` field actually works for cloning, but on the next execution it checks if `git branch` has the tag name - which it doesn't, so it always forces re-cloning the repository.

Since the feature of specifying a commit hash is also missing, I would suggest adding an option for this (named `commit_hash`, or maybe just `commit`), and re-using this option for tag support.
The check on re-execution could work the same in all cases - compare the current `git log` hash with the hash from `git log (options['commit_hash'])`.

As specifying both a branch and a commit seems incompatible to me, those options should probably be mutually exclusive.
(The plugin could verify that the commit is a parent of the given branch's current head, but I don't see the use case for that.)

If that sounds good, and I haven't just overlooked the relevant option somehow, I'd try implementing this tomorrow / over the course of the week.

### PR DESCRIPTION
**Pull request by rohlem**
Fixes #924 .
Direct implementation of what I had in mind, turned out quicker and less painful than I thought.
Works for my use case and shouldn't break anything, though I haven't extensively tested it. (In particular, I'm approaching MTT from 0 and don't have the standard test suite running.)

`Git.py` already has a lot of duplicate lines (f.e. for handling the status of git invocations via shell), my additional copy-pasting didn't help that.
If there's interest I could try factoring them into methods, but for now I kept the style in-line with what it was before.

Sorry that I was too eager to wait for feedback, if there's room left for improving the design I'm willing to rework it into an appropriate alternative.
EDIT: just rebased onto master (not sure how I missed the newest commit).

### CODE REVIEW

**Comment by ribab** [comment]
I will take a look at this for review, although I don't have merge permissions anymore.

Just curious, what brings you to the MTT project @rohlem ?

**Comment by rohlem** [comment]
@ribab Thanks for the look!
The HPC institute at the university I'm attending was looking for an automation/management system for software using MPI, for both manual and automated initiation.
I was initially overwhelmed by the scope of MTT and developed my own solution (source code to-be-released), which is a more free-form system for asynchronous/suspension-aware parameter management (i.e. a poll-able interface that enqueues jobs and exits instead of busy-waiting).

Right now I'm investigating the state of MTT, how it works, and what it might take to implement the process I automated using MTT instead, and/or add suspension with parameter management into MTT.
(At the moment full feature parity seems out-of-scope though, so I'll primarily focus on bug fixes. I might open up a few more issues for tracking, but don't take my work as particularly high-priority or anything.)

**Comment by hppritcha** [comment]
@rohlem would you still like to see this feature in MTT?  I'll check it out if that's the case, and if it works for me go ahead and merge.

**Comment by rohlem** [comment]
@hppritcha I believe this is a generally useful feature that other users could benefit from, so yes.
For example regression testing / bisecting usually wants to work with commit hashes instead of branches.

File: pylib/Tools/Fetch/Git.py [hunk 1]
```diff
@@ -166,17 +168,24 @@ def execute(self, log, keyvals, testDef):
...
         pr = None
         try:
             if cmds['pr'] is not None:
                 pr = cmds['pr']
         except KeyError:
             pass
-        # cannot have both
-        if branch is not None and pr is not None:
+        # cannot have more than one
+        if (1 if branch is not None else 0) + (1 if commit is not None else 0) + (1 if pr is not None else 0) > 1:
```
[the remark this answers is not in the document]
Comment (hppritcha) [reply pylib/Tools/Fetch/Git.py:186]: this is okay for an error check code

**Comment by hppritcha** [comment]
this works for me.  merging.

**Pull request merged**